### DISCUSSION TITLE
Define result of "getResultList" (and others) as never NULL

### DISCUSSION THREAD

**Discussion by Bukama** [jakartaee/persistence#537] [Ideas]
Hey everyone,
this is my first time to get in closer contact with the Jakarta Team and as I can't imagine that this topic was never discussed before and I couldn't find an (open or closed) issue/discussion/etc I decided to ask this question/idea here:

The [Javadoc of the persistence spec of `Query.getResultList()`](https://jakarta.ee/specifications/persistence/3.1/apidocs/jakarta.persistence/jakarta/persistence/query#getResultList()) still does not define what (`NULL` or an empty `List`) is returned when the result set is empty. So the user either depends on the implemention or has to do an null check when using the method.

For example: In old versions of Hibernate the implementation returned `NULL`, newer versions (I can't find since when exactly, but have stopped searching as I looked into the current specs first) return an userfriendly empty list.

So I want to suggest to update the specification that such methods are never allowed to return `null`. I know this might be not backward compatible to older specs, but since Jakarta has begun to release non backward compatible specs I think this is a valid change.

**Comment by gavinking** [comment]
>  In old versions of Hibernate the implementation returned `NULL`

This is simply false. Hibernate always returns an empty list when a query returns no results. And this has been the case since before Hibernate 1.0. No version of Hibernate has ever used `null` to indicate "no results". You're just mistaken about that.

(I suppose it's possible that you once ran into a bug in some very specific exceptional case.)

>  I know this might be not backward compatible to older specs

The expectation has always been that `getResultList()` returns a non-`null` instance of `List`.

So we could certainly make this more explicit, but FWIW I think it's always been pretty clear that an empty list is the correct way to represent "no results". Because using null for this would be obviously insane.

**Comment by gavinking** [comment]
> So we could certainly make this more explicit

See proposed change here:

https://github.com/jakartaee/persistence/pull/538/files

**Comment by Bukama** [comment]
Thank your for your response and your PR!
